### PR TITLE
Support OpenTelemetry tracing

### LINKED ISSUE #28581: Support OpenTelemetry tracing
### Description

For a better observability experience of Keycloak, it'd be great to support a mechanism for obtaining application traces. We'd be able to trace requests with additional information and gather more comprehensive information about the state of the application. We can find performance bottlenecks, determine the cause of application failures in a more agile way, and much more.

The most promising solution for it would be leveraging the open-source OpenTelemetry observability framework, which is also packaged as a Quarkus extension.

https://opentelemetry.io/docs/what-is-opentelemetry/
```[tasklist]
### Tasks
- [ ] https://github.com/keycloak/keycloak/issues/31908
```

### PR DESCRIPTION
**Pull request by mabartos**
Part-of #28581

- [x] New options 
- [x] Options for sampling customization
- [ ] (follow-up) Docs
- [ ] (follow-up) CR integration (at least the endpoint should be included)
- [ ] (follow-up) HTTP client integration

### Jaeger UI outputs (with enabled JDBC tracing):

Screenshot from 2024-04-10 13-51-57

Screenshot from 2024-04-10 13-44-18

Screenshot from 2024-04-10 13-53-07

### CODE REVIEW

**Comment by sschu** [comment]
Thanks for the effort @mabartos! I know @thomasdarimont tried this out and noticed that the Keycloak http client would have to be extended to also add trace headers when making external calls but I don't remember all the details.

**Comment by mabartos** [comment]
@sschu Thanks for the info. It'd be great to gather more feedback on this and know potential improvements.

**Comment by ahus1** [comment]
When the http client doesn't work out of the box first, I would rather want it to be merged and handling the http client in a second PR.

This way we would get feedback as early as possible.

Looking at the screenshots above, this already looks great.

**Comment by ahus1** [comment]
As pointed out by @sschu, sampling of traces is important to keep the overhead and costs at bay. 

Key element is the sampler as configured in [quarkus.otel.traces.sampler](https://quarkus.io/guides/all-config#quarkus-opentelemetry_quarkus-otel-traces-sampler).

The `parentbased_always_on` is a nice default for development environments, still it is not a good default to for production use cases in two aspects: `parentbased` is not a good default for an OIDC service which is exposed to the internet, with no-one filtering out rogue HTTP headers by default from the outside about what should be stored. And `always_on` will log too much for all the requests which wouldn't have that header (which is the default). 

Not sure if we want to differentiate between development setups and production (I usually dislike that as it adds complexity). 

A production setup would usually include:
* Filtering out all HTTP headers from the untrusted internet, so no parent spans can be injected - this would need to be done by a re-encrypt proxy. At least the `tracestate` headers should be filtered, as those would record traces which could swamp a system with a lot of data. 
* If not filtering, don't use the `parentbased_*` tracing at all, which might then be a more sane default for Keycloak.
* Configure Keycloak with `traceidratio` (or if headers from untrusted sources are filtered, with `parentbased_traceidratio`) and a sampler arg that is for example `0.001` to store the traces of one request out of 1000.

Sampling ratio is configured in [quarkus.otel.traces.sampler.arg](https://quarkus.io/guides/all-config#quarkus-opentelemetry_quarkus-otel-traces-sampler-arg).

**Comment by ahus1** [comment]
Once OTEL is enabled, please check if the trace and span IDs are part of the JSON log. I'd even suggest to change the standard logging template to include the trace and span IDs in the log as it improves the analysis of log entries belonging to a request a lot.

**Comment by y-tabata** [comment]
@ahus1 
What else is needed to move this PR forward toward merging?
I see in the comment above that it's good enough for the first PR.

**Comment by ahus1** [comment]
@y-tabata - IMHO "Options for sampling customization" and "Docs" are essential for this to be merged.

**Comment by mabartos** [comment]
JFYI - I plan to work on it this week.

**Comment by mabartos** [comment]
@ahus1 @keycloak/cloud-native 

The majority of the work should be done now and might be ready for some initial review.

As this PR is already quite big, I'd handle docs separately - and other follow-ups. Is it ok?

I also want to investigate more the sampling mechanisms as I didn't check the headers, as @ahus1 pointed out.

**Comment by ahus1** [comment]
@mabartos - thank you for providing this PR. I started it together with Jaeger as outlined in https://quarkus.io/version/3.8/guides/opentelemetry and it looks good to me. 

Compared with the Opentelemetry Java Agent, I saw the different classes for sub-resources, but I don't mind the new view. I also didn't see a Hibernate instrumentation which aggregated things on Entity level first, but only on JDBC level, which is also OK.

The new options `log-console-include-trace`, `log-file-include-trace` and `log-syslog-include-trace` should IMHO not be necessary if the log format automatically adapts once you enable tracing. I saw some code that seems to enable it by default if tracing is enabled, still it didn't show the traces in the log for me unless I started to add `--log-console-include-trace=true`.

I see `tracing-jdbc-enabled` is marked as preview. I didn't find a reference in the Quarkus docs that it is preview. Did I miss something? Would we want to enable it by default if tracing is enabled, as accessing the database is IMHO one of the major sources of latency?

Overall, should we mark the overall tracing as preview while we add more support for HTTP tracing and others?

**Comment by mabartos** [comment]
@ahus1 @vmuzikar @Pepo48 @shawkins 

Should be ready for review with the changed approach described above.

**Review by ahus1 — DISMISSED** [review]
Thank you for this PR. I tested it locally and it looks good to me. I might have skipped the `log-*-include-trace` options and might have forced people to use a manually configured `log-console-format` instead, but let's proceed as you describe it here. 

For the docs, it would be good to have a minimal one-liner in the docs on how to start up Jaeger similar to the [Quarkus docs](https://quarkus.io/guides/opentelemetry#jaeger-to-see-traces-option). Possibly with a Podman instead of docker-compose, it is a single container anyway. 

In the docs, please use extra care to explain: 

* For `tracing-sampler-ratio` that using `1.0` is usually a huge performance overhead in a production environment and you should use a smaller value, and that `tracing-sampler-type` 
* For `tracing-sampler-type` that using `parentbased_traceidratio` is usually the thing you would expect in an environment, but it would require you to trust your callers to enable/disable tracing for a call, and a publicly facing URL with no HTTP header filtering could lead to an overloaded trace store, and that we used `traceidratio` instead for a safer default value.

**Comment by ahus1** [comment]
I didn't merge it as others from the CND team might want to give it a review as well.

**Comment by mabartos** [comment]
@keycloak/cloud-native Hi guys, could you please check it? Thanks!

File: quarkus/config-api/src/main/java/org/keycloak/config/LoggingOptions.java [hunk 1]
```diff
@@ -120,7 +135,13 @@ public String toString() {
     public static final Option<String> LOG_FILE_FORMAT = new OptionBuilder<>("log-file-format", String.class)
             .category(OptionCategory.LOGGING)
             .description("Set a format specific to file log entries.")
-            .defaultValue("%d{yyyy-MM-dd HH:mm:ss,SSS} %-5p [%c] (%t) %s%e%n")
+            .defaultValue(DEFAULT_LOG_FORMAT)
+            .build();
+
+    public static final Option<Boolean> LOG_FILE_INCLUDE_TRACE = new OptionBuilder<>("log-file-include-trace", Boolean.class)
+            .category(OptionCategory.LOGGING)
+            .description(format("Include tracing information to the file log. If '%s' option is specified, this option has no effect.", LOG_FILE_FORMAT.getKey()))
```
Comment (shawkins) [inline quarkus/config-api/src/main/java/org/keycloak/config/LoggingOptions.java:143]:
```suggestion
            .description(format("Include tracing information in the file log. If the '%s' option is specified, this option has no effect.", LOG_FILE_FORMAT.getKey()))
```

File: quarkus/config-api/src/main/java/org/keycloak/config/LoggingOptions.java [hunk 2]
```diff
@@ -158,7 +179,13 @@ public String toString() {
     public static final Option<String> LOG_SYSLOG_FORMAT = new OptionBuilder<>("log-syslog-format", String.class)
             .category(OptionCategory.LOGGING)
             .description("Set a format specific to syslog entries.")
-            .defaultValue("%d{yyyy-MM-dd HH:mm:ss,SSS} %-5p [%c] (%t) %s%e%n")
+            .defaultValue(DEFAULT_LOG_FORMAT)
+            .build();
+
+    public static final Option<Boolean> LOG_SYSLOG_INCLUDE_TRACE = new OptionBuilder<>("log-syslog-include-trace", Boolean.class)
+            .category(OptionCategory.LOGGING)
+            .description(format("Include tracing information to the syslog. If '%s' option is specified, this option has no effect.", LOG_SYSLOG_FORMAT.getKey()))
```
Comment (shawkins) [inline quarkus/config-api/src/main/java/org/keycloak/config/LoggingOptions.java:187]:
```suggestion
            .description(format("Include tracing information in the syslog. If the '%s' option is specified, this option has no effect.", LOG_SYSLOG_FORMAT.getKey()))
```

**Review by shawkins — DISMISSED** [review]
LGTM, just some possible minor tweaks to the descriptions.

**Review by ahus1 — APPROVED** [review]
Thank you for this PR. I approve it based on my preview review and looking at the help texts. Leaving it for open for others to review it as well.

**Review by shawkins — APPROVED** [review]
LGTM, seems like a nice enhancement @mabartos

**Pull request merged**